### PR TITLE
refactor(all): unify method signatures across backends

### PR DESCRIPTION
**Pull request by gforsyth**
We want to ensure that, for a given backend, that the argument names, plus usage of positional, positional-only, keyword, and keyword-only arguments match, so that there is API consistency when moving between backends.

I've grabbed a few small parts of some of the utilities in Scott Sanderson's
`python-interface` project (https://github.com/ssanderson/python-interface).

While the upstream is no longer maintained, the goal of that project
aligns quite well with some of the issues we face with maintaining
consistent interfaces across backends.

Whether we keep the `pytest` approach for testing the signatures or not, this has proven reasonably effective at finding spots where our argument names aren't matching.  (I think I'm in favor of keeping it, since it flagged up some inconsistencies in `to_parquet_dir` that were added after I opened this PR initially)

I've tried to keep the various refactor commits reasonably atomic so that this isn't an absolute nightmare to review.

### CODE REVIEW

File: ibis/backends/tests/test_signatures.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+@pytest.mark.parametrize("base_cls, method", params)
+def test_signatures(base_cls, method, backend_cls):
+    if not hasattr(backend_cls, method):
+        pytest.skip(f"Method {method} not present in {backend_cls}, skipping...")
+
+    base_sig = inspect.signature(getattr(base_cls, method))
+    backend_sig = inspect.signature(getattr(backend_cls, method))
+
+    assert compatible(base_sig, backend_sig, check_annotations=False)
```
Comment (gforsyth) [inline ibis/backends/tests/test_signatures.py:26]: I imagine we could break up the base classes into groups where we think type annotations should match vs. those where we expect some variation.

Comment (gforsyth) [inline ibis/backends/tests/test_signatures.py:26]: I have these backwards here, and might make all of these (required) keyword arguments?

Usage is `compatible(implementation, interface_definition, ...)`

**Comment by gforsyth** [comment]
xref #8994

**Comment by ncclementi** [comment]
xref: https://github.com/ibis-project/ibis/issues/9125

File: ibis/backends/__init__.py [hunk 2]
```diff
@@ -477,7 +478,7 @@ def to_parquet(
     def to_parquet_dir(
         self,
         expr: ir.Table,
-        directory: str | Path,
+        path: str | Path,
```
Comment (ncclementi) [inline ibis/backends/__init__.py:481]: The use of `directory` was an ask in the `to_parquet_dir` PR, see https://github.com/ibis-project/ibis/pull/9781#discussion_r1709931796 . DO we want to go backwards on that?

**Comment by gforsyth** [comment]
Closing in favor of #10008

**Pull request closed without merging**